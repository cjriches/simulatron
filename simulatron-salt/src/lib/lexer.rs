use logos::Logos;

#[derive(Logos, Debug, PartialEq, Eq, Copy, Clone)]
pub enum TokenType {
    // Keywords
    #[token("const")]
    Const,
    #[token("static")]
    Static,
    #[token("mut")]
    Mut,
    #[token("byte")]
    Byte,
    #[token("half")]
    Half,
    #[token("word")]
    Word,

    // Punctuation
    #[token("[")]
    OpenSquare,
    #[token("]")]
    CloseSquare,
    #[token(",")]
    Comma,
    #[token(":")]
    Colon,

    // Literal components
    #[regex(r"-?([0-9]+(e-?[0-9]+)?|0b[01]+|0x[A-Fa-f0-9]+)")]
    IntLiteral,
    #[regex(r"-?[0-9]+\.[0-9]+(e-?[0-9]+)?")]
    FloatLiteral,
    #[regex(r"'(\\[^\n]|[^\n\\'])'")]
    CharLiteral,
    #[regex(r#""(\\[^\n]|[^\n\\"])*""#)]
    StringLiteral,

    // Identifiers
    #[regex(r"[A-Za-z_][A-Za-z0-9_]*")]
    Identifier,

    // Comments
    #[regex(r"//[^\n]*")]
    Comment,

    // Whitespace
    #[regex(r"\r|\n|\r\n")]
    Newline,
    #[regex(r"[^\S\n\r]+")]
    Whitespace,

    // Unrecognised tokens.
    #[error]
    Unknown,
}

#[derive(Debug, Clone)]
pub struct Token<'a> {
    pub tt: TokenType,
    pub span: logos::Span,
    pub slice: &'a str,
}

/// Wrap the Logos implementation with some extra buffering.
pub struct Lexer<'a> {
    inner: logos::Lexer<'a, TokenType>,
    pushed_back: Vec<Token<'a>>,
}

impl<'a> Lexer<'a> {
    /// Create a new token stream from the given source.
    pub fn new(source: &'a str) -> Self {
        Self {
            inner: TokenType::lexer(source),
            pushed_back: Vec::with_capacity(1),
        }
    }

    /// Push a token back onto the front of the stream.
    pub fn push_back(&mut self, token: Token<'a>) {
        self.pushed_back.push(token)
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = Token<'a>;

    /// Get the next token.
    fn next(&mut self) -> Option<Self::Item> {
        // Use the pushed back tokens first.
        if !self.pushed_back.is_empty() {
            return self.pushed_back.pop();
        }

        // Otherwise, get from the lexer.
        return match self.inner.next() {
            Some(tt) => {
                Some(Token {
                    tt,
                    span: self.inner.span(),
                    slice: self.inner.slice(),
                })
            },
            None => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_snapshot;
    use std::fmt::Write;

    fn assert_tokens_snapshot(path: &str) {
        let input = std::fs::read_to_string(path).unwrap();
        let mut lexer = Lexer::new(&input);
        let mut output = String::new();
        while let Some(t) = lexer.next() {
            if let TokenType::Newline = t.tt {
                writeln!(output, "{:?}", t.tt).unwrap();
            } else {
                writeln!(output, "{:?} `{}`", t.tt, t.slice).unwrap();
            }
        }
        assert_snapshot!(output);
    }

    /// Test the simplest possible input: a single instruction.
    #[test]
    fn test_minimal() {
        assert_tokens_snapshot("examples/minimal.simasm");
    }

    /// Test a small instruction block.
    #[test]
    fn test_instruction_block() {
        assert_tokens_snapshot("examples/instruction-block.simasm");
    }

    /// Test character literals.
    #[test]
    fn test_char_literal() {
        assert_tokens_snapshot("examples/char-literal.simasm");
    }

    /// Test string literals.
    #[test]
    fn test_string_literal() {
        assert_tokens_snapshot("examples/string-literal.simasm");
    }

    /// Test numeric literals
    #[test]
    fn test_numeric_literals() {
        assert_tokens_snapshot("examples/numeric-literals.simasm");
    }

    /// Test comments.
    #[test]
    fn test_comments() {
        assert_tokens_snapshot("examples/comments.simasm");
    }

    /// Test a simple hello world program.
    #[test]
    fn test_hello_world() {
        assert_tokens_snapshot("examples/hello-world.simasm");
    }

    /// Test a program full of invalid tokens.
    #[test]
    fn test_bad_tokens() {
        assert_tokens_snapshot("examples/bad-tokens.simasm");
    }
}
