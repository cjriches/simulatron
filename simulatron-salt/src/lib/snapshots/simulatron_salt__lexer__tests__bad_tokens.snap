---
source: src/lib/lexer.rs
expression: output

---
Identifier `This`
Whitespace ` `
Identifier `is`
Whitespace ` `
Identifier `_really_`
Whitespace ` `
Identifier `wrong`
Whitespace ` `
Identifier `for`
Whitespace ` `
Identifier `a`
Whitespace ` `
Unknown `*`
Period `.`
Identifier `simasm`
Whitespace ` `
Identifier `file`
Period `.`
Whitespace ` `
Identifier `There`
Whitespace ` `
Identifier `are`
Whitespace ` `
Identifier `all`
Whitespace ` `
Identifier `sorts`
Whitespace ` `
Identifier `of`
Whitespace ` `
Identifier `incorrect`
Whitespace ` `
Identifier `tokens`
Comma `,`
Whitespace ` `
Identifier `generating`
Whitespace ` `
Identifier `lots`
Whitespace ` `
Identifier `of`
Whitespace ` `
Identifier `errors`
Period `.`
Newline
Newline
Identifier `int`
Whitespace ` `
Identifier `main`
Unknown `(`
Identifier `int`
Whitespace ` `
Identifier `argc`
Comma `,`
Whitespace ` `
Identifier `char`
Whitespace ` `
Unknown `*`
Unknown `*`
Identifier `argv`
Unknown `)`
Newline
Unknown `{`
Newline
Whitespace `    `
Unknown `/`
Unknown `*`
Whitespace ` `
Identifier `Don`
Unknown `'`
Identifier `t`
Whitespace ` `
Identifier `you`
Whitespace ` `
Identifier `love`
Whitespace ` `
Identifier `this`
Whitespace ` `
Identifier `comment`
Whitespace ` `
Identifier `style`
Unknown `?`
Whitespace ` `
Unknown `*`
Unknown `/`
Newline
Whitespace `    `
Identifier `printf`
Unknown `(`
StringLiteral `"Hello, World!"`
Unknown `)`
Unknown `;`
Newline
Unknown `}`
Newline
Newline
Identifier `main`
Whitespace ` `
Colon `:`
Colon `:`
Whitespace ` `
Identifier `IO`
Whitespace ` `
Unknown `(`
Unknown `)`
Newline
Identifier `main`
Whitespace ` `
Unknown `=`
Whitespace ` `
Identifier `putStrLn`
Whitespace ` `
StringLiteral `"Hello, World!"`
Whitespace `  `
Minus `-`
Minus `-`
Whitespace ` `
Identifier `We`
Unknown `'`
Identifier `re`
Whitespace ` `
Identifier `writing`
Whitespace ` `
Identifier `Haskell`
Comma `,`
Whitespace ` `
Identifier `right`
Unknown `?`
Newline
Newline
Identifier `if`
Whitespace ` `
Identifier `__name__`
Whitespace ` `
Unknown `=`
Unknown `=`
Whitespace ` `
StringLiteral `"__main__"`
Colon `:`
Newline
Whitespace `    `
Identifier `print`
Unknown `(`
StringLiteral `"Hello, world!"`
Unknown `)`
Whitespace `  `
Unknown `#`
Whitespace ` `
Identifier `Say`
Whitespace ` `
Identifier `hello`
Period `.`
Newline
Newline
Comment `// The next bit is almost correct!`
Newline
Const `const`
Whitespace ` `
Identifier `YEET`
Whitespace ` `
Unknown `(`
DecLiteral `42`
Unknown `)`
Newline

